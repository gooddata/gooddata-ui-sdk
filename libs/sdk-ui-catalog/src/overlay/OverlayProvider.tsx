// (C) 2025 GoodData Corporation
import React, { type ReactNode } from "react";

import { OverlayController, OverlayControllerProvider } from "@gooddata/sdk-ui-kit";

// Error message in AD is 3100 and loader is 3200. We need to be above both
// The chart tooltip is overriden to be 3305, otherwise it's not visible

const overlayController = OverlayController.getInstance(3300);

export function OverlayProvider({ children }: { children: ReactNode }) {
    return (
        <OverlayControllerProvider overlayController={overlayController}>
            {children}
        </OverlayControllerProvider>
    );
}
