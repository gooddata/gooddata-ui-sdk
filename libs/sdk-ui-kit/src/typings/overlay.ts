// (C) 2020-2025 GoodData Corporation
import { IAlignPoint } from "./positioning.js";
import { IRegion } from "./domUtilities.js";

/**
 * @internal
 */

export type GetOptimalAlignment = {
    targetRegion: IRegion;
    selfRegion: IRegion;
    ignoreScrollOffsets?: boolean;
    alignPoints: IAlignPoint[];
    getViewportRegion?: (ignoreScrollOffsets: boolean) => void;
    getDocumentRegion?: () => void;
};

/**
 * @internal
 */

export type GetPositionedSelfRegion = {
    targetRegion: IRegion;
    selfRegion: IRegion;
    bodyRegion: IRegion;
    alignPoint: IAlignPoint;
};

/**
 * @internal
 */

export type GetOptimalAlignmentForRegion = {
    boundaryRegion: any;
    targetRegion: any;
    selfRegion: any;
    alignPoints: IAlignPoint[];
};

/**
 * @internal
 */

export type Alignment = {
    left: number;
    top: number;
    right: number;
    width?: number;
    height?: number;
    align: string;
};

/**
 * @internal
 */
export interface IOptimalAlignment {
    alignment: Alignment;
    visiblePart?: number;
}

/**
 * @internal
 */
export type SameAsTargetPosition = "sameAsTarget";

/**
 * @internal
 */
export type OverlayPositionType = "absolute" | "fixed" | SameAsTargetPosition;
