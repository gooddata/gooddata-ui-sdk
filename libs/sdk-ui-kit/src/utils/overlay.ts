// (C) 2007-2023 GoodData Corporation
import mapValues from "lodash/mapValues.js";
import merge from "lodash/merge.js";
import { elementRegion } from "./domUtilities.js";
import {
    GetOptimalAlignment,
    GetPositionedSelfRegion,
    GetOptimalAlignmentForRegion,
    IOptimalAlignment,
} from "../typings/overlay.js";
import { IOffset, IAlignPoint } from "../typings/positioning.js";
import { IRegion } from "../typings/domUtilities.js";

const ALIGN_RATIOS_X: Record<string, number> = {
    l: 0,
    r: 1,
    c: 0.5,
};

const ALIGN_RATIOS_Y: Record<string, number> = {
    t: 0,
    b: 1,
    c: 0.5,
};

const DEFAULT_OFFSET_VALUES = {
    x: 0,
    y: 0,
};

export const DEFAULT_ALIGN_POINTS = [
    {
        align: "cc cc",
        offset: {
            x: 0,
            y: 0,
        },
    },
];

const FULLY_HIDDEN_ALIGNMENT = {
    alignment: {
        left: 0,
        top: 0,
        right: 0,
        align: DEFAULT_ALIGN_POINTS[0].align,
    },
    visiblePart: -1,
};

function getDefaultViewportRegion(ignoreScrollOffsets: boolean) {
    const region: IRegion = elementRegion(document.body);

    return {
        top: ignoreScrollOffsets ? 0 : document.documentElement.scrollTop,
        left: ignoreScrollOffsets ? 0 : document.documentElement.scrollLeft,
        width: region.width,
        height: region.height,
    };
}

function getDefaultDocumentRegion() {
    return {
        top: 0,
        left: 0,
        height: document.documentElement.scrollHeight,
        width: document.documentElement.scrollWidth,
    };
}

function getAlignRatios(coordinatesString: string) {
    if (!(coordinatesString[0] in ALIGN_RATIOS_Y) || !(coordinatesString[1] in ALIGN_RATIOS_X)) {
        throw new Error(`Invalid point specification: ${coordinatesString}`);
    }

    return {
        x: ALIGN_RATIOS_X[coordinatesString[1]],
        y: ALIGN_RATIOS_Y[coordinatesString[0]],
    };
}

function getAlignPointRatios(alignPointAlign: string) {
    const alignPointPositions = alignPointAlign.split(" ");

    return {
        target: getAlignRatios(alignPointPositions[0]),
        self: getAlignRatios(alignPointPositions[1]),
    };
}

function getRegionOffset(region: IRegion, ratio: IOffset, direction: number = 1) {
    return {
        x: region.width * ratio.x * direction,
        y: region.height * ratio.y * direction,
    };
}

function getAlignPointOffset(alignPointOffset: IOffset) {
    return merge({}, DEFAULT_OFFSET_VALUES, alignPointOffset);
}

function addOffset(position: IOffset, offset: IOffset) {
    return {
        x: position.x + offset.x,
        y: position.y + offset.y,
    };
}
function subtractOffset(position: IOffset, offset: IOffset) {
    return {
        x: position.x - offset.x,
        y: position.y - offset.y,
    };
}

function getGlobalOffset(target: IRegion) {
    return {
        x: target.left,
        y: target.top,
    };
}

/**
 * Get left-top coordinates of a child summing up offsets of a child,
 * align points and left-top coordinates of target region
 */
function getGlobalPosition(
    bodyRegion: IRegion,
    targetRegion: IRegion,
    selfRegion: IRegion,
    alignPoint: IAlignPoint,
) {
    const alignRatio = getAlignPointRatios(alignPoint.align);

    const targetRegionOffset = getRegionOffset(targetRegion, alignRatio.target);
    const selfRegionOffset = getRegionOffset(selfRegion, alignRatio.self, -1);
    const alignPointOffset = getAlignPointOffset(alignPoint.offset);
    const cumulativeOffset = addOffset(alignPointOffset, addOffset(targetRegionOffset, selfRegionOffset));

    const globalOffset = getGlobalOffset(targetRegion);
    const finalOffset = addOffset(globalOffset, cumulativeOffset);

    return subtractOffset(finalOffset, getGlobalOffset(bodyRegion));
}

/**
 * Move region to position specified by anchor point,
 * using its left-top coordinates
 */
function moveRegionToPosition(region: IRegion, anchorPoint: IOffset) {
    return merge({}, region, {
        left: anchorPoint.x,
        top: anchorPoint.y,
        width: region.width,
        height: region.height,
    });
}

/**
 * Move self region to new position using its left-top coordinates
 */
function getPositionedSelfRegion({
    targetRegion,
    selfRegion,
    bodyRegion,
    alignPoint,
}: GetPositionedSelfRegion) {
    const globalPosition = getGlobalPosition(bodyRegion, targetRegion, selfRegion, alignPoint);

    return moveRegionToPosition(selfRegion, globalPosition);
}

/**
 * Get fraction of visible area
 */
function getRatioOfVisibleRegion(boundaryRegion: IRegion, region: IRegion) {
    const maximalVisibleArea = region.width * region.height;

    if (maximalVisibleArea === 0) {
        return 0;
    }

    const leftBorder = Math.max(boundaryRegion.left, region.left);
    const rightBorder = Math.min(boundaryRegion.left + boundaryRegion.width, region.left + region.width);

    const topBorder = Math.max(boundaryRegion.top, region.top);
    const bottomBorder = Math.min(boundaryRegion.top + boundaryRegion.height, region.top + region.height);

    const visibleWidth = Math.max(rightBorder - leftBorder, 0);
    const visibleHeight = Math.max(bottomBorder - topBorder, 0);

    const visibleArea = visibleWidth * visibleHeight;
    const visibleRatio = visibleArea / maximalVisibleArea;

    return Math.max(0, visibleRatio);
}

function isFullyVisible(alignment: IOptimalAlignment) {
    return alignment.visiblePart === 1;
}

function isFullyHidden(alignment: IOptimalAlignment) {
    return alignment.visiblePart === 0;
}

/**
 * Calculate most visible alignment of self region which
 * would be positioned to target region using specified
 * align points
 */
export function getOptimalAlignmentForRegion({
    boundaryRegion,
    targetRegion,
    selfRegion,
    alignPoints,
}: GetOptimalAlignmentForRegion): IOptimalAlignment {
    const bodyRegion: IRegion = elementRegion(document.body);
    let mostVisibleAlignment = FULLY_HIDDEN_ALIGNMENT;

    for (let i = 0; i < alignPoints.length; i += 1) {
        const positionedSelfRegion = getPositionedSelfRegion({
            alignPoint: alignPoints[i],
            targetRegion,
            selfRegion,
            bodyRegion,
        });

        const visiblePart = getRatioOfVisibleRegion(boundaryRegion, positionedSelfRegion);

        if (visiblePart > mostVisibleAlignment.visiblePart) {
            const coordinates = {
                left: positionedSelfRegion.left,
                top: positionedSelfRegion.top,
                right:
                    boundaryRegion.left + boundaryRegion.width - positionedSelfRegion.left - selfRegion.width,
            };

            mostVisibleAlignment = {
                alignment: merge(mapValues(coordinates, Math.floor), {
                    align: alignPoints[i].align,
                }),
                visiblePart,
            };
        }

        if (isFullyVisible(mostVisibleAlignment)) {
            return mostVisibleAlignment;
        }
    }

    return mostVisibleAlignment;
}

/**
 * Calculate optimal alignment of self region
 * using viewport boundaries. Try to position
 * using body boundaries if the region
 * is fully hidden in the viewport
 */
export function getOptimalAlignment({
    targetRegion,
    selfRegion,
    ignoreScrollOffsets,
    alignPoints,
    getViewportRegion = getDefaultViewportRegion,
    getDocumentRegion = getDefaultDocumentRegion,
}: GetOptimalAlignment): IOptimalAlignment {
    let optimalAlign = getOptimalAlignmentForRegion({
        boundaryRegion: getViewportRegion(ignoreScrollOffsets),
        targetRegion,
        selfRegion,
        alignPoints,
    });

    if (isFullyHidden(optimalAlign)) {
        optimalAlign = getOptimalAlignmentForRegion({
            boundaryRegion: getDocumentRegion(),
            targetRegion,
            selfRegion,
            alignPoints,
        });
    }

    return optimalAlign;
}

export function getOverlayStyles() {
    const styles = window.getComputedStyle(document.body);
    const marginTop = parseFloat(styles.marginTop) || 0;
    const marginBottom = parseFloat(styles.marginBottom) || 0;

    const html = {
        position: "relative",
        height: "100%",
    };
    const body = {
        position: "relative",
        height: "100%",
    };

    if (marginTop || marginBottom) {
        body.height = `calc(100% - ${marginTop + marginBottom}px)`;
    }

    return {
        html,
        body,
    };
}
