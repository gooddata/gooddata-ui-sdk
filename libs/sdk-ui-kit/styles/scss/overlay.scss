// (C) 2007-2024 GoodData Corporation
@use "mixins";
@use "typo-mixins";
@use "variables";

.gd-dropdown.overlay {
    // reset overlays padding
    padding: 0;
}

.overlay-wrapper {
    z-index: 5001;
}

.overlay {
    @include typo-mixins.gd-paragraph;

    position: relative;
    padding: 0.1px; // Prevent collapsing margins
    border-radius: 3px;
    font-family: variables.$gd-font-primary;
    background-color: variables.$gd-color-white;

    @include mixins.box-shadow(0 1px 20px var(--gd-shadow-color-from-theme, rgba(20, 56, 93, 0.2)));

    @media #{variables.$small-only} {
        border-radius: 0;
    }
}

.gd-overlay-content {
    @include mixins.default-styles();
}

.gd-error-overlay-content {
    .modalityPlugin-mask-visible {
        z-index: 10001;
        opacity: 0.95;
    }
}

.gd-error-overlay {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    text-align: center;

    .gd-error-overlay-icon {
        margin-bottom: 20px;
    }

    h2 {
        margin-bottom: 10px;
    }

    .gd-error-overlay-text {
        line-height: 23px;
        color: variables.$gd-color-text;
        font-size: 14px;
    }

    .gd-error-overlay-button {
        margin-top: 20px;
    }
}
