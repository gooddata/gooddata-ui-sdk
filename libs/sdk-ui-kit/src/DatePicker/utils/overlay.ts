// (C) 2007-2020 GoodData Corporation
import mapValues from "lodash/mapValues";
import merge from "lodash/merge";
import { elementRegion } from "../../utils/domUtilities";
import {
    Region,
    GetOptimalAlignment,
    GetPositionedSelfRegion,
    GetOptimalAlignmentForRegion,
} from "../typings";
import { IOffset, IAlignPoint } from "../../typings/positioning";

const ALIGN_RATIOS_X = {
    l: 0,
    r: 1,
    c: 0.5,
};

const ALIGN_RATIOS_Y = {
    t: 0,
    b: 1,
    c: 0.5,
};

const DEFAULT_OFFSET_VALUES = {
    x: 0,
    y: 0,
};

export const DEFAULT_ALIGN_POINTS = [
    {
        align: "cc cc",
        offset: {
            x: 0,
            y: 0,
        },
    },
];

const FULLY_HIDDEN_ALIGNMENT = {
    alignment: {
        left: 0,
        top: 0,
        right: 0,
        align: DEFAULT_ALIGN_POINTS[0].align,
    },
    visiblePart: -1,
};

function getDefaultViewportRegion(ignoreScrollOffsets: boolean) {
    const region: any = elementRegion(document.body);

    return {
        top: ignoreScrollOffsets ? 0 : document.documentElement.scrollTop,
        left: ignoreScrollOffsets ? 0 : document.documentElement.scrollLeft,
        width: region.width,
        height: region.height,
    };
}

function getDefaultDocumentRegion() {
    return {
        top: 0,
        left: 0,
        height: document.documentElement.scrollHeight,
        width: document.documentElement.scrollWidth,
    };
}

function getAlignRatios(coordinatesString: string) {
    if (!(coordinatesString[0] in ALIGN_RATIOS_Y) || !(coordinatesString[1] in ALIGN_RATIOS_X)) {
        throw new Error(`Invalid point specification: ${coordinatesString}`);
    }

    return {
        x: ALIGN_RATIOS_X[coordinatesString[1]],
        y: ALIGN_RATIOS_Y[coordinatesString[0]],
    };
}

function getAlignPointRatios(alignPointAlign: string) {
    const alignPointPositions = alignPointAlign.split(" ");

    return {
        target: getAlignRatios(alignPointPositions[0]),
        self: getAlignRatios(alignPointPositions[1]),
    };
}

function getRegionOffset(region: Region, ratio: IOffset, direction: number = 1) {
    return {
        x: region.width * ratio.x * direction,
        y: region.height * ratio.y * direction,
    };
}

function getAlignPointOffset(alignPointOffset: IOffset) {
    return merge({}, DEFAULT_OFFSET_VALUES, alignPointOffset);
}

function addOffset(position: IOffset, offset: IOffset) {
    return {
        x: position.x + offset.x,
        y: position.y + offset.y,
    };
}

function getGlobalOffset(target: Region) {
    return {
        x: target.left,
        y: target.top,
    };
}

/**
 * Get left-top coordinates of a child summing up offsets of a child,
 * align points and left-top coordinates of target region
 */
function getGlobalPosition(targetRegion: Region, selfRegion: Region, alignPoint: IAlignPoint) {
    const alignRatio = getAlignPointRatios(alignPoint.align);

    const targetRegionOffset = getRegionOffset(targetRegion, alignRatio.target);
    const selfRegionOffset = getRegionOffset(selfRegion, alignRatio.self, -1);
    const alignPointOffset = getAlignPointOffset(alignPoint.offset);
    const cumulativeOffset = addOffset(alignPointOffset, addOffset(targetRegionOffset, selfRegionOffset));

    const globalOffset = getGlobalOffset(targetRegion);

    return addOffset(globalOffset, cumulativeOffset);
}

/**
 * Move region to position specified by anchor point,
 * using its left-top coordinates
 */
function moveRegionToPosition(region: Region, anchorPoint: IOffset) {
    return merge({}, region, {
        left: anchorPoint.x,
        top: anchorPoint.y,
        width: region.width,
        height: region.height,
    });
}

/**
 * Move self region to new position using its left-top coordinates
 */
function getPositionedSelfRegion({ targetRegion, selfRegion, alignPoint }: GetPositionedSelfRegion) {
    const globalPosition = getGlobalPosition(targetRegion, selfRegion, alignPoint);

    return moveRegionToPosition(selfRegion, globalPosition);
}

/**
 * Get fraction of visible area
 */
function getRatioOfVisibleRegion(boundaryRegion: Region, region: Region) {
    const maximalVisibleArea = region.width * region.height;

    if (maximalVisibleArea === 0) {
        return 0;
    }

    const leftBorder = Math.max(boundaryRegion.left, region.left);
    const rightBorder = Math.min(boundaryRegion.left + boundaryRegion.width, region.left + region.width);

    const topBorder = Math.max(boundaryRegion.top, region.top);
    const bottomBorder = Math.min(boundaryRegion.top + boundaryRegion.height, region.top + region.height);

    const visibleWidth = Math.max(rightBorder - leftBorder, 0);
    const visibleHeight = Math.max(bottomBorder - topBorder, 0);

    const visibleArea = visibleWidth * visibleHeight;
    const visibleRatio = visibleArea / maximalVisibleArea;

    return Math.max(0, visibleRatio);
}

function isFullyVisible(alignment: any) {
    return alignment.visiblePart === 1;
}

function isFullyHidden(alignment: any) {
    return alignment.visiblePart === 0;
}

/**
 * Calculate most visible alignment of self region which
 * would be positioned to target region using specified
 * align points
 */
export function getOptimalAlignmentForRegion({
    boundaryRegion,
    targetRegion,
    selfRegion,
    alignPoints,
}: GetOptimalAlignmentForRegion) {
    let mostVisibleAlignment = FULLY_HIDDEN_ALIGNMENT;

    for (let i = 0; i < alignPoints.length; i += 1) {
        const positionedSelfRegion = getPositionedSelfRegion({
            alignPoint: alignPoints[i],
            targetRegion,
            selfRegion,
        });

        const visiblePart = getRatioOfVisibleRegion(boundaryRegion, positionedSelfRegion);

        if (visiblePart > mostVisibleAlignment.visiblePart) {
            const coordinates = {
                left: positionedSelfRegion.left,
                top: positionedSelfRegion.top,
                right:
                    boundaryRegion.left + boundaryRegion.width - positionedSelfRegion.left - selfRegion.width,
            };

            mostVisibleAlignment = {
                alignment: merge(mapValues(coordinates, Math.floor), {
                    align: alignPoints[i].align,
                }),
                visiblePart,
            };
        }

        if (isFullyVisible(mostVisibleAlignment)) {
            return mostVisibleAlignment;
        }
    }

    return mostVisibleAlignment;
}

/**
 * Calculate optimal alignment of self region
 * using viewport boundaries. Try to position
 * using body boundaries if the region
 * is fully hidden in the viewport
 */
export function getOptimalAlignment({
    targetRegion,
    selfRegion,
    ignoreScrollOffsets,
    alignPoints,
    getViewportRegion = getDefaultViewportRegion,
    getDocumentRegion = getDefaultDocumentRegion,
}: GetOptimalAlignment) {
    let optimalAlign = getOptimalAlignmentForRegion({
        boundaryRegion: getViewportRegion(ignoreScrollOffsets),
        targetRegion,
        selfRegion,
        alignPoints,
    });

    if (isFullyHidden(optimalAlign)) {
        optimalAlign = getOptimalAlignmentForRegion({
            boundaryRegion: getDocumentRegion(),
            targetRegion,
            selfRegion,
            alignPoints,
        });
    }

    return optimalAlign;
}
