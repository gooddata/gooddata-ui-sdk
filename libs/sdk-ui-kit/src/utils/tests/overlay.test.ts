// (C) 2020-2022 GoodData Corporation
import { getOptimalAlignment, getOptimalAlignmentForRegion } from "../overlay";

function generateOverlay() {
    const boundaryRegion = {
        top: 100,
        left: 100,
        width: 1000,
        height: 1000,
    };

    const targetRegion = {
        top: 900,
        left: 900,
        width: 100,
        height: 100,
    };

    const selfRegion = {
        top: 0,
        left: 0,
        width: 200,
        height: 200,
    };

    const alignPoints = [
        {
            align: "tr tl",
        },
        {
            align: "tl tr",
        },
    ];

    return {
        boundaryRegion,
        targetRegion,
        selfRegion,
        alignPoints,
    };
}

function generateOverlayWithCustomAlign(align: string) {
    const boundaryRegion = {
        top: 0,
        left: 0,
        width: 1000,
        height: 1000,
    };

    const targetRegion = {
        top: 50,
        left: 50,
        width: 100,
        height: 100,
    };

    const selfRegion = {
        top: 0,
        left: 0,
        width: 200,
        height: 200,
    };

    const alignPoints = [
        {
            align,
        },
    ];

    return {
        boundaryRegion,
        targetRegion,
        selfRegion,
        alignPoints,
    };
}

describe("Overlay utils", () => {
    it("should align inside document when aligned element is outside of the viewport", () => {
        const { alignPoints, targetRegion, selfRegion } = generateOverlay();

        const getViewportRegion = () => ({
            left: 0,
            top: 0,
            width: 500,
            height: 500,
        });

        const getDocumentRegion = () => ({
            left: 0,
            top: 0,
            width: 1000,
            height: 2500,
        });

        const optimalAlignment = getOptimalAlignment({
            targetRegion,
            selfRegion,
            alignPoints,
            getViewportRegion,
            getDocumentRegion,
        });

        expect(optimalAlignment).toMatchSnapshot();
    });

    describe("without offset", () => {
        it("should use second alignPoint because it would be hidden or partially visible using the first one", () => {
            const { boundaryRegion, targetRegion, selfRegion, alignPoints } = generateOverlay();
            const optimalAlignment = getOptimalAlignmentForRegion({
                boundaryRegion,
                targetRegion,
                selfRegion,
                alignPoints,
                overlayRootElement: document.body,
            });

            expect(optimalAlignment).toMatchSnapshot();
        });

        it("should use first alignPoint because aligned node fits fully into viewport", () => {
            const { boundaryRegion, targetRegion, alignPoints } = generateOverlay();

            const smallerSelfRegion = {
                width: 100,
                height: 100,
                top: 0,
                left: 0,
            };

            const optimalAlignPoint = getOptimalAlignmentForRegion({
                boundaryRegion,
                targetRegion,
                selfRegion: smallerSelfRegion,
                alignPoints,
                overlayRootElement: document.body,
            });

            expect(optimalAlignPoint).toEqual({
                alignment: {
                    left: 1000,
                    right: 0,
                    top: 900,
                    align: "tr tl",
                },
                visiblePart: 1,
            });
        });

        it("should use best result when node is not fully visible in any case", () => {
            const { boundaryRegion, targetRegion, alignPoints } = generateOverlay();

            const biggerSelfRegion = {
                width: 500,
                height: 500,
                top: 0,
                left: 0,
            };

            const optimalAlignPoint = getOptimalAlignmentForRegion({
                boundaryRegion,
                targetRegion,
                selfRegion: biggerSelfRegion,
                alignPoints,
                overlayRootElement: document.body,
            });

            expect(optimalAlignPoint).toEqual({
                alignment: {
                    left: 400,
                    right: 200,
                    top: 900,
                    align: "tl tr",
                },
                visiblePart: 0.4,
            });
        });
    });

    describe("with offset", () => {
        const alignPointsWithOffset = [
            {
                align: "tr tl",
                offset: {
                    x: -100,
                    y: -100,
                },
            },
            {
                align: "tl tr",
                offset: {
                    x: 0,
                    y: -100,
                },
            },
        ];

        it("should use first alignPoint because aligned node fits fully into viewport thanks to negative offset", () => {
            const { boundaryRegion, targetRegion, selfRegion } = generateOverlay();
            const optimalAlignPoint = getOptimalAlignmentForRegion({
                boundaryRegion,
                targetRegion,
                selfRegion,
                alignPoints: alignPointsWithOffset,
                overlayRootElement: document.body,
            });

            expect(optimalAlignPoint).toEqual({
                alignment: {
                    left: 900,
                    right: 0,
                    top: 800,
                    align: "tr tl",
                },
                visiblePart: 1,
            });
        });

        it("should use second alignPoint because aligned node does not fit into viewport using first alignPoint", () => {
            const { boundaryRegion, targetRegion } = generateOverlay();
            const biggerSelfRegion = {
                width: 300,
                height: 300,
                top: 0,
                left: 0,
            };

            const optimalAlignPoint = getOptimalAlignmentForRegion({
                boundaryRegion,
                targetRegion,
                selfRegion: biggerSelfRegion,
                alignPoints: alignPointsWithOffset,
                overlayRootElement: document.body,
            });

            expect(optimalAlignPoint).toEqual({
                alignment: {
                    left: 600,
                    right: 200,
                    top: 800,
                    align: "tl tr",
                },
                visiblePart: 1,
            });
        });
    });

    describe("Alignment computing", () => {
        it("should align overlay to given node top-left both", () => {
            const align = "tl tl";

            const { boundaryRegion, targetRegion, selfRegion, alignPoints } =
                generateOverlayWithCustomAlign(align);
            const optimalAlignment = getOptimalAlignmentForRegion({
                boundaryRegion,
                targetRegion,
                selfRegion,
                alignPoints,
                overlayRootElement: document.body,
            });

            expect(optimalAlignment).toMatchSnapshot();
        });

        it("should align overlay top left to given node center-center", () => {
            const align = "cc tl";

            const { boundaryRegion, targetRegion, selfRegion, alignPoints } =
                generateOverlayWithCustomAlign(align);
            const optimalAlignment = getOptimalAlignmentForRegion({
                boundaryRegion,
                targetRegion,
                selfRegion,
                alignPoints,
                overlayRootElement: document.body,
            });

            expect(optimalAlignment).toMatchSnapshot();
        });

        it("should align overlay center-center to given node bottom right", () => {
            const align = "br cc";

            const { boundaryRegion, targetRegion, selfRegion, alignPoints } =
                generateOverlayWithCustomAlign(align);
            const optimalAlignment = getOptimalAlignmentForRegion({
                boundaryRegion,
                targetRegion,
                selfRegion,
                alignPoints,
                overlayRootElement: document.body,
            });

            expect(optimalAlignment).toMatchSnapshot();
        });

        it("should align overlay bottom-left to given node bottom-right", () => {
            const align = "br bl";

            const { boundaryRegion, targetRegion, selfRegion, alignPoints } =
                generateOverlayWithCustomAlign(align);
            const optimalAlignment = getOptimalAlignmentForRegion({
                boundaryRegion,
                targetRegion,
                selfRegion,
                alignPoints,
                overlayRootElement: document.body,
            });

            expect(optimalAlignment).toMatchSnapshot();
        });
    });
});
