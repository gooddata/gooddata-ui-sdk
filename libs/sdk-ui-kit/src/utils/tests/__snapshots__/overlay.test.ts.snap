// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Overlay utils > Alignment computing > should align overlay bottom-left to given node bottom-right 1`] = `
{
  "alignment": {
    "align": "br bl",
    "left": 150,
    "right": 650,
    "top": -50,
  },
  "visiblePart": 0.75,
}
`;

exports[`Overlay utils > Alignment computing > should align overlay center-center to given node bottom right 1`] = `
{
  "alignment": {
    "align": "br cc",
    "left": 50,
    "right": 750,
    "top": 50,
  },
  "visiblePart": 1,
}
`;

exports[`Overlay utils > Alignment computing > should align overlay to given node top-left both 1`] = `
{
  "alignment": {
    "align": "tl tl",
    "left": 50,
    "right": 750,
    "top": 50,
  },
  "visiblePart": 1,
}
`;

exports[`Overlay utils > Alignment computing > should align overlay top left to given node center-center 1`] = `
{
  "alignment": {
    "align": "cc tl",
    "left": 100,
    "right": 700,
    "top": 100,
  },
  "visiblePart": 1,
}
`;

exports[`Overlay utils > should align inside document when aligned element is outside of the viewport 1`] = `
{
  "alignment": {
    "align": "tl tr",
    "left": 700,
    "right": 100,
    "top": 900,
  },
  "visiblePart": 1,
}
`;

exports[`Overlay utils > without offset > should use second alignPoint because it would be hidden or partially visible using the first one 1`] = `
{
  "alignment": {
    "align": "tl tr",
    "left": 700,
    "right": 200,
    "top": 900,
  },
  "visiblePart": 1,
}
`;
